Fix #5486 & part of #5343: Introducing new wiki page for code coverage usage and limitations (#5483)
oppia/oppia-android: wiki/Writing-tests-with-good-behavioural-coverage.md

}
```

**Line Coverage Result:** This test covers the line of code but doesn’t test all scenarios.

### Behavioural Coverage Testing

To ensure behavioral coverage, the test needs to verify various conditions:

```
@Test fun testGetName_withDefaultValue_result(getName()) {
    // Default value when no name is provided
    assertThat(getName()).isEqualTo(" ") 
}

@Test fun testGetName_withNullName_throwsException() {
    // Exception for null value
    assertThrows<IllegalArgumentException> { getName(null) } 
}

@Test fun testGetName_withSpecialCharacters_throwsException() {
    // Exception for special characters
    assertThrows<IllegalArgumentException> { getName("!@#$%^&*()") } 
}

@Test fun testGetName_withEmptyName_result(getName("")) {
    // Empty string should use default value
    assertThat(getName("")).isEqualTo(" ") 
}

@Test fun testGetName_withWhitespaceName_result(getName("   ")) {
    // Whitespace name
    assertThat(getName("   ")).isEqualTo("   ") 
}
```

**Behavioural Coverage Result:** These tests ensure that all potential behaviors, including edge cases and exceptions, are covered, providing a more thorough validation.

> ### Quality > Percentage

Reviewer: I suggest also adding something to this to include:

Lines covered is more helpfully used as an indicator for when a behavioral analysis of testing should be done. This means that:
1. 100% coverage doesn't guarantee all important behaviors are tested.
2. <100% coverage *does* guarantee that at least one important behavior is missing from testing.
3. The goal when improving code coverage should be to test all behaviors that can be determined as related to the lines not covered in tests (and also analyze existing behaviors in case other obvious missing cases can be tested, too, even if their corresponding lines are already executed in tests).

(Might need some rewording, but I think the concept is valuable to communicate explicitly).

Author: Thanks for outlining them clearly. I made just a slight expansion and added it to the section. - https://github.com/oppia/oppia-android/blob/b1a5bd76878cfe21ce66ca01df984d0d904c4337/wiki/Writing-tests-with-good-behavioural-coverage.md#quality--percentage

- Use descriptive names for your test methods.

Naming Convention:
```
testAction_withOneCondition_withSecondCondition_hasExpectedOutcome
```

Example:
```
testCheckSign_forPositiveInput_returnsPositive()
testCheckSign_forNegativeInput_returnsNegative()
testCheckSign_forZeroInput_returnsNeitherPositiveNorNegative()
```

## 3. Focusing on Specific Test Cases

When testing a function that performs multiple actions, it's crucial to write test cases that focus on individual aspects of the function. This approach allows you to isolate and identify issues more effectively if a test fails, rather than dealing with a more complex failure analysis.

### Why Focus on Specific Test Cases?
Testing one thing at a time helps:

- Identify Issues More Easily: If a test fails, you know precisely which aspect of the function is problematic.
- Improve Test Clarity: Each test case has a clear purpose, making tests easier to read and maintain.
- Isolate Failures: Helps in pinpointing issues related to a specific behavior or output.

Consider a function that manages a food order process. This function does the following:

1. Lists the food items.
2. Calculates the total price.

> 3. Display the order.

Reviewer: ```suggestion
3. Displays the order.
```

For consistency (and the correct verb form for functions/behaviors).

Author: Thanks, updated it.

4. Checks if the payment has been made and provides the corresponding message.

```
fun processOrder(order: List<String>, paymentMade: Boolean): String {
    // List the food items
    val itemList = order.joinToString(", ")
    
    // Calculate total price (mocked here for simplicity)
    val totalPrice = order.size * 10.0
    
    // Display order
    println("Order: $itemList")
    println("Total: $totalPrice")
    
    // Payment status
    return if (paymentMade) "Payment successful" else "Payment pending"
}
```

**Potential output with payment made:**

```
Order: Pizza, Burger
Total: 20.0
Result: Payment successful
```

### Testing the Entire Functionality

**Single Test Case:**

```
@Test
fun testProcessOrder_allSteps_returnsCorrectMessage() {
    val result = processOrder(listOf("Pizza", "Burger"), paymentMade = true)
    assertThat(result).isEqualTo("Payment successful")
}
```

**Difficulties in Testing All Aspects Together:**

- Complex Failure Diagnosis: If this test fails, you need to diagnose whether the issue lies in listing items, calculating the total, displaying the order, or payment status.

> - Less Focused: It does not target individual aspects of the function, making it harder to identify which specific action failed.

Reviewer: ```suggestion
- Less Focused: It does not target individual aspects of the function, making it harder to identify which specific action failed for cases when the test overall is failing.
```

For a bit more clarity.

Author: Thanks, updated as suggested.

### Testing Specific Aspects

**Test Case 1: Testing Listing items**

```
@Test

> fun testProcessOrder_listsItems_correctly() {

Reviewer: Here & below: these test names aren't actually following the suggested naming format mentioned above very clearly. :) Can they be made more clear? Happy to think about it and offer suggestions on clearer names, if needed.

Ditto for all example test names being introduced in this PR (more than just these 4 need some work, I think).

Author: Yeah, should've rushed these.
 
Updated them with temporary names. Would love to hear your suggestions for more precise naming and update them later. - https://github.com/oppia/oppia-android/blob/1c1531d9d938dcee6c47078a1349d2ef2762de27/wiki/Writing-tests-with-good-behavioural-coverage.md#testing-specific-aspects

Will also check the rest of the sections to see if anything else deviates from convention.

  assertThat(evaluateAccess(30, false)).isEqualTo("Membership required")
}

@Test
fun testEvaluateAccess_minorMember() {
  assertThat(evaluateAccess(16, true)).isEqualTo("Access denied")
}

@Test
fun testEvaluateAccess_minorNonMember() {
  assertThat(evaluateAccess(15, false)).isEqualTo("Access denied")
}
```

Testing all branches and conditions ensures that your function can handle all possible scenarios, making it more reliable and easier to maintain.

## 6. Exception and Error Handling

**Exception Handling:**

Exceptions are unexpected conditions or events that disrupt the normal flow of a program. They handle situations not part of the regular program flow, like invalid user input or file operation issues.

Examples:
- `NullPointerException` (accessing a null reference)
- `IllegalArgumentException` (invalid function argument)
- `IOException` (I/O operation failure)

This allows to catch and handle issues, allowing the program to continue running or fail in a controlled manner.

For instance, to handle division by zero error.

```
fun divide(a: Int, b: Int): Int {
    if (b == 0) throw ArithmeticException("Division by zero")
    return a / b
}
```

> **Error Handling:**

Reviewer: Drop this part and mention how to handle domain errors, instead. Runtime errors (as in java.lang.Error) are outside the scope of usual testing since their boundary conditions are often either impossible to deterministically set up, or impractical (such as for OOM).

By 'errors' in my earlier comment I meant domain errors. For example: if a function can fail, but recover, it may print an error message and/or return a failure status. This is a failure in the sense that something went wrong, but it's not _exceptional_ in that execution needs to cease because the application is now in an irrecoverable state. It's really important to test both of these types of failure modes since they are both designed into implementations.

Author: Understood it based on the parsing error examples mentioned in the discussion and used those exact samples to present the sections. - https://github.com/oppia/oppia-android/blob/b1a5bd76878cfe21ce66ca01df984d0d904c4337/wiki/Writing-tests-with-good-behavioural-coverage.md#6-exception-and-error-handling

```
@Test
fun testDivideNumbers_forValidInputs_returnsExpectedResult() {
    val result = divideNumbers(10, 2)
    assertThat(result).isEqualTo(5)
}
```

### Testing Edge Cases

Test edge cases where exceptions might be thrown, such as boundary values or extreme input scenarios.

**Function with Edge Case Handling:**

```
fun calculateDiscount(price: Double, discountPercent: Double): Double {
    if (price < 0 || discountPercent < 0) { 
        throw IllegalArgumentException("Price and discount cannot be negative")
    }
    return price - (price * discountPercent / 100)
}
```

**Test Case:**

```
@Test
fun testCalculateDiscount_forNegativePrice_throwsIllegalArgumentException() {
    val exception = assertThrows<IllegalArgumentException> {
        calculateDiscount(-100.0, 10.0)
    }
    assertThat(exception.message).contains("Price and discount cannot be negative")
}

@Test
fun testCalculateDiscount_forNegativeDiscount_throwsIllegalArgumentException() {
    val exception = assertThrows<IllegalArgumentException> {
        calculateDiscount(100.0, -10.0)
    }
    assertThat(exception.message).contains("Price and discount cannot be negative")
}
```

> # Guidelines for Testing Public API

Reviewer: ```suggestion
# Guidelines for Testing Public APIs
```

Content-wise: I think this section is currently serving as a collection of useful examples to demonstrate different aspects of testing public APIs, but it doesn't quite explain the process of _how_ to, which I think is especially useful to attempt since then we're basically teaching people the concepts and methodologies they need in order to incorporate these lessons.

From my perspective, I like to analyze the public API of a class into a list of behaviors that I'll write tests for (I actually usually write out these behaviors in the form of test names using the format you mentioned above, but that's just a personal preference). I _only_ look at this list because, using the test format above, I can very easily see different edge cases that may need to be considered or that I may have missed by looking at the list of behaviors I've written holistically and cross-referencing them with the public API that I'm analyzing.

I don't even start writing tests until this list of behaviors is complete.

I'm wondering if being strongly prescriptive on this approach would be more helpful to readers, especially if they don't yet have a methodology for writing tests based on a public API?

Author: Ah, thanks for explaining your approach to writing these tests; that’s super helpful to know in general and for organizing the section. I’ve added a concise version based on how you laid it out. Would like to know if there are any additional steps or details that might be needed. - https://github.com/oppia/oppia-android/blob/b1a5bd76878cfe21ce66ca01df984d0d904c4337/wiki/Writing-tests-with-good-behavioural-coverage.md#testing-public-api

    val unavailableFile = File("nonExistentFile.pdf") // File that does not exist

    val exception = assertThrows<IllegalArgumentException> {
        account.withdraw("user", "password", 200.0, file = unavailableFile)
    }
    assertThat(exception).contains("File not found")
}
```

c. Testing with Valid File

```
@Test
fun testWithdraw_withValidFile_processesFile() {
    val account = BankAccount(1000.0, "user", "password")
    val validFile = File("passbook.pdf")
    validFile.createNewFile() // Ensure the file exists for the test

    val output = ByteArrayOutputStream()
    System.setOut(PrintStream(output))

    account.withdraw("user", "password", 200.0, file = validFile)

    assertThat(output.toString().trim()).isEqualTo("Processing file: passbook.pdf")
    System.setOut(System.out)
}
```

These tests collectively cover critical aspects of file handling, ensuring robust and reliable API functionality.

### 7. Testing Absence of Unwanted Output

In addition to validating correct handling of valid and invalid files, it's also important to ensure that unwanted output or behavior does not occur.

**Test:**

a. Ensure No Processing Message for Missing File

> ```

Reviewer: Here & everywhere else: Kotlin examples can be highlighted using '```kotlin' for better syntax highlighting.

Author: Sure, updated every kotlin snippet to use ```kotlin. Thanks.

    System.setOut(PrintStream(output))

    account.withdraw("user", "password", 200.0)

    assertThat(output.toString().trim()).contains("New balance is 800.0")
    System.setOut(System.out)
}
```

**e. To verify with the balance is updated correctly:**

Ensures that the balance is updated correctly after a withdrawal.

```
@Test
fun testWithdraw_withSufficientBalance_updatesBalanceCorrectly() {
    account.withdraw("user", "password", 200.0)
    assertThat(account.balance).isEqualTo(800.0)
}
```

These tests cover various aspects of the withdrawal functionality, ensuring that the balance updates correctly with receipts, passbooks and output messages. Although the core functionality of withdrawing funds and updating the balance is consistent, it can be observed in multiple ways. Each test focuses on a specific verification method while ultimately validating the same core functionality.

## Testing Practices for Effective Validation

The following best practices are meant to assist in the organization and execution of tests. By following these guidelines, you can ensure that your testing process is efficient, thorough, and well-structured.

> ### 1. Use of Helper Functions

Reviewer: Important detail to include here (and maybe as part of the setup/teardown parts): **only** details not directly relevant to the _behavior_ of a test should be moved to helpers or the before/after functions. These aren't details that are necessary for the test to succeed (I assume all lines are necessary, otherwise they shouldn't be included). Instead, these are details that, if missing from the main test body, would make it confusing as a reader to verify that the implementation of a test actually matches the behavior it's trying to verify.

This is sort of the art form of writing tests: the more unrelated details that can be extracted, the better since it helps keep the test much more focused.

It's maybe also important to clarify somewhere that we use tests to verify implementation code is correct, but humans verify that test code is correct. For this reason, we err on the side of code duplication in tests (unlike production code) or other forms of code repetition since clarity is more important than conciseness in test writing.

Author: Understood, but the two approaches seem contradictory to how I had them earlier, (making it not so fitting in the after before annotations) so restructured it to be

### Structuring Test Functionalities 
- When and How to Divide Responsibilities
  - Helper Functions
  - @Before and @After Annotations
- When Not to Divide Responsibilities
  - Prioritizing Clarity Over Conciseness in Tests
- Importance of Descriptive Test Names

https://github.com/oppia/oppia-android/blob/1c1531d9d938dcee6c47078a1349d2ef2762de27/wiki/Writing-tests-with-good-behavioural-coverage.md#structuring-test-functionalities

will revise as per feedback.

Reviewer: I actually think these (seemingly opposing) sections are fine since they're covering specific cases (in both situations) when tests should and should not be split up. It actually seems rather clear to me.

Helper functions are crucial for avoiding repetitive code, especially for pre-setup tasks that are used across multiple test cases. In this scenario, initializing the bank data or ensuring the proper state of the BankAccount object can be encapsulated in helper functions.

```
// Helper function to initialize a BankAccount with default data
fun createDefaultBankAccount(): BankAccount {

>     return BankAccount(1000.0, "user", "password")

Reviewer: Here & elsewhere: please use 2 space tabbing for consistency with the Oppia Android codebase.

Author: Done, updated every snippet to use 2-space tabbing.

}
```

This helps to reduce redundancy and maintain consistency across multiple test cases by centralizing repetitive setup tasks.

### 2. Setup and Teardown using `@Before` and `@After`

Using `@Before` and `@After` annotations ensures that common setup and cleanup tasks are automatically executed before and after each test case, maintaining a clean and consistent test environment.

**Example:**

```
class BankAccountTests {

    private lateinit var account: BankAccount

    @Before
    fun setUp() {
        // Initialize a BankAccount instance before each test
        account = createDefaultBankAccount()
    }

    @After

>     fun cleanup() {

Reviewer: Nit: cleanUp (verb form), but perhaps tearDown is better given the section name?

Author: Yeah, while naming it as `restoreStreams` was followed for restoring system outputs in various places, 

```
@After
fun restoreStreams() {
  System.setOut(originalOut)
}
```

thought `cleanUp` would provide a more generalized name, but `tearDown` is the standard practice across the codebase, will stick with that and updated it to use `tearDown`. Thanks.

Edit:
Oh, just realized I gave the section name as 'tearDown'. My bad.

    // Verify that attempting to use a non-existent file results in an exception
    val exception = assertThrows<IllegalArgumentException> {
        account.withdraw("user", "password", 200.0, file = unavailableFile)
    }
    assertThat(exception.message).contains("File not found")
}

@Test
fun testWithdraw_withDefaultReceipt_noReceiptPrinted() {
    val output = ByteArrayOutputStream()
    System.setOut(PrintStream(output))

    account.withdraw("user", "password", 200.0)

    // Ensure that no receipt is printed when the default receipt flag is used
    assertThat(output.toString().trim()).doesNotContain("Receipt:")
    System.setOut(System.out)
}
```

**Importance of Specific Naming and Conditions:**

- Clarity: Specific names and conditions make tests easier to understand and manage.
- Focus: Helps pinpoint exact scenarios being tested, improving test coverage.
- Debugging: Clear names and conditions aid in quickly identifying the cause of failures.
- Documentation: Serves as self-documentation, providing insight into test purpose and scope.
- Maintenance: Simplifies updates and modifications by clearly defining what each test covers.

> # How to Map a Line of Code to Its Corresponding Behaviors?

Reviewer: ```suggestion
# How to Map a Line of Code to Its Corresponding Behaviors
```

The other section headers aren't questions.

Author: Okay. Updated as suggested.

> ## Table of Contents

Reviewer: I think this is an excellent start. Some other things that I think are worth noting:
- We aim for 100% behavioral coverage, not just line of code coverage.
- The concept of a class's 'public API' (that is, its public members and how they behave in different contexts or with different call values/states).
- How to map a line of code to its corresponding behaviors (i.e. mapping it back to its public API and the behaviors that can lead to it being called). This is a _very_ useful thing to write up since it allows us to provide developers with a basic guide on how to effectively improve code coverage when writing tests.

Perhaps it's also worth framing the naming style that you repeat in this page in terms of behaviors: tests ought to verify a single behavior, and a behavior is:
- Tied to a single action (like a method call).
- Requires specific state to occur (its preconditions).
- Can be observed in a number of ways, but the test only needs to verify it in a single way (the expected outcome).

Author: Thank you so much for providing a detailed breakdown of the page requirements and specific sections. That was a beacon.

- I’ve added the behavioral section to address the behavioral coverage requirements, and the writing effective tests have been updated to align with these requirements - https://github.com/oppia/oppia-android/blob/51393a49069bb14832eaebdba6178695f103437c/wiki/Writing-tests-with-good-behavioural-coverage.md#writing-effective-tests . 

- For the public API, I’ve included a sample general function to illustrate the scenario and demonstrate various ways to test and map it. However, I’m unsure about all possible states specific to the Oppia codebase. - https://github.com/oppia/oppia-android/blob/51393a49069bb14832eaebdba6178695f103437c/wiki/Writing-tests-with-good-behavioural-coverage.md#guidelines-for-testing-public-api

- Added the "How to map a line of Code" section and with the mapping section, I used the same public API to explain a basic scenario. Should there be any additional complex scenarios that should be included, particularly those specific to the codebase? - https://github.com/oppia/oppia-android/blob/51393a49069bb14832eaebdba6178695f103437c/wiki/Writing-tests-with-good-behavioural-coverage.md#how-to-map-a-line-of-code-to-its-corresponding-behaviors

- I’ve also added the name framing to the testing practices section - https://github.com/oppia/oppia-android/blob/51393a49069bb14832eaebdba6178695f103437c/wiki/Writing-tests-with-good-behavioural-coverage.md#3-descriptive-test-names 

- Included tests to illustrate the need for specific conditions. -  https://github.com/oppia/oppia-android/blob/51393a49069bb14832eaebdba6178695f103437c/wiki/Writing-tests-with-good-behavioural-coverage.md#3-focusing-on-specific-test-cases

- Also, added the section to explain how a single outcome can be tested in different ways. - https://github.com/oppia/oppia-android/blob/51393a49069bb14832eaebdba6178695f103437c/wiki/Writing-tests-with-good-behavioural-coverage.md#8-testing-a-single-outcome-in-multiple-ways

Reviewer: Thanks! This is a really excellent set of changes and I've gone through, reviewed them, and commented.

Re: public API, I actually have a different perspective on how to think about this (behavior first rather than state based). See my newer comment on this specific concept.

Re: mapping to line of code, I actually think your example is really good. I don't see a need for a more complex example at this time.

> - [Writing Tests with Good Behavioral Coverage](##writing-tests-with-good-behavioral-coverage)

Reviewer: Extra '#' in the anchor needs to be removed.

Author: Thanks. I’ve made that correction in the revised table of contents.

- [Why is Behavioral Coverage Necessary?](#why-is-behavioral-coverage-necessary)
- [Writing Effective Tests](#writing-effective-tests)
    - [Understand the Requirements](#1-understanding-the-requirements)
    - [Writing Clear and Descriptive Test Cases](#2-writing-clear-and-descriptive-test-cases)
    - [Covering Different Scenarios](#3-covering-different-scenarios)
    - [Exception Handling](#4-exception-handling)

# Writing Tests with Good Behavioral Coverage

> Writing tests with good behavioral coverage involves ensuring that your code is tested thoroughly to meet its requirements and behave correctly in various scenarios. This approach aims to verify not only that your code works but also that it handles different inputs and edge cases as expected.

Reviewer: There's another element to this worth pointing out: we specifically want to focus on behavioral coverage over line coverage since the same line of source code is typically used for many different behaviors, so verifying that the line is called at least once only serves to verify (in the worst case) a single behavior which allows the other behaviors to be broken without automated test detection (despite having reportedly high line coverage).

Author: Thank you so much that gave me a much better understanding of what should be actually included and deciding the structure and flow of the page. 

Included this section to distinguish both practices - https://github.com/oppia/oppia-android/blob/51393a49069bb14832eaebdba6178695f103437c/wiki/Writing-tests-with-good-behavioural-coverage.md#understanding-behavioral-coverage-and-its-importance

# Why is Behavioral Coverage Necessary?

Behavioral coverage is crucial because it helps ensure that your code functions correctly under all expected conditions, not just the typical use cases. By covering various scenarios, you can:

- Detect Bugs Early: Identify and fix issues before they affect users.
- Improve Code Quality: Ensure that the code meets its requirements and handles edge cases.
- Maintain Robustness: Ensure that changes or additions to the codebase do not introduce new issues.

> For more details on testing methodologies specific to Oppia Android, please refer to the Oppia Testing Wiki.

Reviewer: Please add a link to the specific page.

Author: Done, added them with the revised version references.

> For more details on testing methodologies specific to Oppia Android, please refer to the Oppia Testing Wiki.

Reviewer: Please add a link to the specific page.

Author: Done, added them.

# Writing Effective Tests

## 1. Understanding the Requirements

Before writing tests thoroughly review documentation, user stories, or specifications for the functionality you're testing.

> **Example:**

Reviewer: Needs a newline above this if you want it in its own paragraph.

Author: Followed the suggestion with the revised version.

If you're testing a function that checks if a number is positive or negative, make sure it returns "Positive" for numbers greater than or equal to zero and "Negative" for numbers less than zero.

```
@Test
fun testCheckSign_forPositiveInput_returnsPositive() {
    assertEquals("Positive", checkSign(5))
}

@Test
fun testCheckNumber_forNegativeInput_returnsNegative() {
    assertEquals("Negative", checkSign(-3))
}
```

# 2. Writing Clear and Descriptive Test Cases

Each test case should:

- Clearly describe the scenario and expected outcome.
- Use descriptive names for your test methods.

Naming Convention:
```
testAction_withOneCondition_withSecondCondition_hasExpectedOutcome
```

Example:
```
testCheckSign_forPositiveInput_returnsPositive()
testCheckSign_forNegativeInput_returnsNegative()
testCheckSign_forZeroInput_returnsNeitherPositiveNorNegative()
```

## 3. Covering Different Scenarios

> Ensure tests cover:

Reviewer: Also: extreme cases (when appropriate). In your example code, it would be good to to test large value integers, too, to provide confidence that the function isn't overly tailored to the select inputs of the tests (this is a possible side effect from TDD/BDD development strategies).

Author: Would this section: https://github.com/oppia/oppia-android/blob/51393a49069bb14832eaebdba6178695f103437c/wiki/Writing-tests-with-good-behavioural-coverage.md#testing-different-scenerios, covering MAX and MIN boundaries be appropriate in this case?

Reviewer: Sure I think that seems reasonable.

- Positive Cases: Valid inputs that should pass.
- Negative Cases: Invalid inputs or edge cases.
- Boundary Cases: Inputs at the edge of acceptable ranges.

```
@Test
fun testCheckSign_forZeroInput_returnsNeitherPositiveNorNegative() {
  assertEquals("Neither Positive Nor Negative", checkSign(0)) // Boundary value
}

@Test
fun testCheckSign_forPositiveBoundaryInput_returnsPositive() {
  assertEquals("Positive", checkSign(1)) // Just above the boundary value
}

@Test
fun testCheckSign_forNegativeBoundaryInput_returnsNegative() {
  assertEquals("Negative", checkSign(-1)) // Just below the boundary value
}
```

> ## 4. Exception Handling

Reviewer: Suggest generalizing this to be both exception and error handling, since they are similar in the Java/Kotlin world but notably different.

Broadly, we want to make sure that failures are correctly tested. Either that means verifying a specific state or return value indicating failure occurs, or it means catching and verifying an exception.

For your reference, exceptions ought to only be thrown in exceptional (i.e. irrecoverable) cases, but that's not necessarily important in this context (since errors of all forms ought to be tested).

Author: Updated it with describing both - https://github.com/oppia/oppia-android/blob/51393a49069bb14832eaebdba6178695f103437c/wiki/Writing-tests-with-good-behavioural-coverage.md#6-exception-and-error-handling, while covering cases with exceptions. Is this right? And would it require any additional tests, or should the "Testing Exceptions" title be renamed to "Testing Exceptions and Errors"?

Reviewer: I added a new comment to discuss this in more detail, actually. Let me know if you have any further questions or need clarity on that thread--closing this one since a lot has been changed since my original comment.

Exception Handling is a critical aspect of testing that ensures your code can handle and recover from error conditions gracefully.

- Ensure Error Conditions Are Managed Gracefully:
    - You need to verify that your application responds correctly when invalid or unexpected inputs are provided. For instance, if your function is designed to throw an exception when it receives null, you should test that it does so correctly.

- Verify Correct Exception Types:
    - Testing should confirm that the correct type of exception is thrown. This ensures that your error handling logic is specific and accurate.

```
@Test
fun testCheckSign_forNullInput_throwsIllegalArgumentException() {
    assertThrows<IllegalArgumentException> {
        checkSign(null)
    }
}
```

- Check for Proper Exception Messages:

>     - It's important to ensure that exceptions include meaningful and informative messages that help diagnose issues. This can be tested by verifying the content of the exception message.

Reviewer: Not only this, but checking the error message is important when generic exceptions (like IllegalStateException) are thrown to ensure the _correct_ occurrence of the exception is being thrown when expected.

Author: Does it mean to check the actual error message with generic exceptions (eg: as mentioned IllegalStateException) to have a descriptive message. So instead of having a generic message like

```
throw IllegalStateException("Error") // Generic message
```

but to have a descriptive message

```
throw IllegalStateException("Cannot perform action: User account is inactive.") // Specific message
```

And does this cover this case? - https://github.com/oppia/oppia-android/blob/51393a49069bb14832eaebdba6178695f103437c/wiki/Writing-tests-with-good-behavioural-coverage.md#checking-for-proper-exception-messages or should I add something to actually differentiate what a generic message would (which is a don'ts) and specify the need for specific message (which is a dos). 

Will confirm my understanding on this and update as per feedback.

Reviewer: I think your example touches on precisely what I was thinking, but it could maybe benefit from an additional note on why we care about verifying the message. While the existing reasoning is correct (that the message is clear), it's also especially necessary for generic exceptions to ensure the *correct* occurrence of that exception is being thrown. Consider a function that throws IllegalStateExceptions in two places (which is quite common since we use ``check`` in a bunch of places)--not verifying the error message could mean accidentally verifying the wrong exception is thrown which makes the test less useful as it's not actually verifying what it's meant to.

Author: (missed responding to this earlier) - Thanks for providing more context. I’ve added a section on the Importance of Specific Error Messages, based on the same reasoning: https://github.com/oppia/oppia-android/blob/a4c48d6dec31934b35fa4defafc113ad0c4f3661/wiki/Writing-tests-with-good-behavioural-coverage.md#importance-of-specific-error-messages

## Understanding Behavioral Coverage and its Importance

Focusing on behavioral coverage is essential because:

- Multiple Behaviors per Line: A single line of code can produce multiple behaviors depending on input values and conditions.
- Comprehensive Testing: High line coverage doesn’t guarantee all behaviors are tested. Behavioral coverage ensures every scenario, including edge cases and exceptions, is validated.
- Quality Over Quantity: Achieving high line coverage might give a false sense of security. Testing all possible behaviors ensures higher quality and robustness.

Let's understand it better with a sample function. Consider this function that validates a name:

```kotlin
fun getName(name: String? = " ") =
  name?.takeIf { it.all { char -> char.isLetterOrWhitespace() } }
    ?: throw IllegalArgumentException("Invalid name")
```

### Line Coverage Testing

A basic test case for line coverage might look like this:

```kotlin

> @Test fun testValidName() {

Reviewer: Nit here & everywhere: let's make sure '@Test' is on its own line like we do in regular tests for consistency (even though it will make the page slightly longer--consistency is important here, I think, since this is an authoritative guide).

Author: Got it! moved every @Test annotation and its function to their own lines.

  - [Covering Different Scenarios](#4-covering-different-scenarios)
  - [Covering All Branches, Paths, and Conditions](#5-covering-all-branches-paths-and-conditions)
  - [Exception and Error Handling](#6-exception-and-error-handling)
  - [Absence of Unwanted Output](#7-absence-of-unwanted-output)
- [Testing Public API](#testing-public-api)

> - [Structuring Test Functionalities](#structuring-test-functionalities)

Reviewer: Suggest instead: 'Structuring Test Bodies'

Author: Done, updated 'functionalities' to 'bodies'.

Let's understand it better with a sample function. Consider this function that validates a name:

```kotlin
fun getName(name: String? = " ") =
  name?.takeIf { it.all { char -> char.isLetterOrWhitespace() } }
    ?: throw IllegalArgumentException("Invalid name")
```

### Line Coverage Testing

A basic test case for line coverage might look like this:

```kotlin
@Test fun testValidName() {
  // Tests line coverage by hitting the line where name is accessed

>   assertThat(getName("Alice"), equalTo("Alice")) 

Reviewer: Here & elsewhere: seems like this got changed incorrectly? Should still be ``isEqualTo`` for the ``assertThat``s.

Author: Yeah, I was moving things in and out with GitHub Markdown formatting and Android Studio. I should have accidently moved in an older version (will also cross check other updates I did since then), but I’ve now changed, `equalTo` to use `.isEqualTo`. Thanks!

While line coverage might reach 100% with a single test, it doesn’t ensure that all scenarios are tested. Behavioral coverage ensures quality by validating all possible scenarios and edge cases, which is more important for reliable software.

**Evaluation and Enhancement Flow:**

> ```sh

Reviewer: Nit: not shell code (can just be unhighlighted code text, I think).

Author: Yeah, kept them as they popped some colors for the illustration, but it looks neat in monochrome as well. I've removed the highlighting now.

Author: Replaced this textual illustration with mermaid blocks.

Reviewer: Ah I see. Well fortunately the diagrams seem much cleaner, so no need for extra syntax highlighting I think!

**Single Test Case:**

```kotlin
@Test
fun testProcessOrder_allSteps_returnsCorrectMessage() {
  val result = processOrder(listOf("Pizza", "Burger"), paymentMade = true)
  assertThat(result).isEqualTo("Payment successful")
}
```

**Difficulties in Testing All Aspects Together:**

- Complex Failure Diagnosis: If this test fails, you need to diagnose whether the issue lies in listing items, calculating the total, displaying the order, or payment status.
- Less Focused: It does not target individual aspects of the function, making it harder to identify which specific action failed for cases when the test overall is failing.

### Testing Specific Aspects

**Test Case 1: Testing Listing items**

```kotlin
@Test
fun testProcessOrder_providedWithList_displaysListOfItems() {
  processOrder(listOf("Pizza", "Burger"), paymentMade = true)
  val output = outContent.toString().trim()
  assertThat(output, containsString("Order: Pizza, Burger"))
}
```

**Test Case 2: Calculates Total**

```kotlin
@Test
fun testProcessOrder_forListItems_calculatesCorrectTotalPrice() {
  processOrder(listOf("Pizza", "Burger"), paymentMade = true)
  val output = outContent.toString().trim()

>   assertThat(output, containsString("Total: 20.0"))

Reviewer: Here & elsewhere: should be ``assertThat(output).contains("Total: 20.0")``

Author: Done, updated with contains.

  if (denominator == 0) throw IllegalArgumentException("Denominator cannot be zero")
  return numerator / denominator
}
```

In this example, if the denominator is zero, an `IllegalArgumentException` is thrown. This is a standard way to handle situations where continuing the program as usual doesn’t make sense due to an error in the input.

**Testing Exceptions:**

The primary focus when testing exceptions is ensuring that the correct exception is thrown in the appropriate circumstances and that the program can handle it gracefully.

**Test:**

```kotlin
@Test
fun testDivideNumbers_forZeroDenominator_throwsIllegalArgumentException() {
  val exception = assertThrows<IllegalArgumentException> {
    divideNumbers(10, 0)
  }
  assertThat(exception).hasMessageThat().contains("Denominator cannot be zero")
}
```

This test verifies that the divideNumbers function throws an IllegalArgumentException when the denominator is zero. It also checks that the exception message contains the expected text. Testing exceptions involves confirming that the application correctly identifies and handles these unexpected situations.

### Domain Errors

> Domain errors are errors related to the business logic or rules of the application, rather than technical issues like those covered by exceptions. These errors occur when the data or conditions within the domain do not meet the specific rules or constraints defined by the business. Domain errors are expected conditions and are handled within the normal flow of the application, rather than through exceptions.

Reviewer: ```suggestion
Domain errors are errors related to the business logic or rules of the application, rather than technical issues like those covered by exceptions. These errors occur when the data or conditions within the domain do not meet the specific rules or constraints defined during application design. Domain errors are expected conditions and are handled within the normal flow of the application, rather than through exceptions.
```

'business' reads slightly oddly to me here, so reworded to hopefully improve understandability.

Author: Okay, updated it to - "Domain errors are errors related to **the logic** or rules of the application, rather than technical issues like those covered by exceptions. These errors occur when the data or conditions within the domain do not meet the specific rules or constraints **defined during application design**. Domain errors are expected conditions and are handled within the normal flow of the application, rather than through exceptions."

Reviewer: Ah much cleaner--thanks!

```kotlin
fun getSubmitTimeError(text: String): FractionParsingError {
  if (text.isNullOrBlank()) {
    return FractionParsingError.EMPTY_INPUT
  }
  if (invalidCharsLengthRegex.find(text) != null) {
    return FractionParsingError.NUMBER_TOO_LONG
  }
  if (text.endsWith("/")) {
    return FractionParsingError.INVALID_FORMAT
  }
  val fraction = parseFraction(text)
  return when {
    fraction == null -> FractionParsingError.INVALID_FORMAT
    fraction.denominator == 0 -> FractionParsingError.DIVISION_BY_ZERO
    else -> FractionParsingError.VALID
  }
}
```

This function checks various conditions on the input string text to determine whether it meets the criteria for a valid fraction. Each condition that fails returns a specific domain error from the FractionParsingError enum. Unlike exceptions, these errors are expected as part of the application's logic and represent specific, recoverable conditions that the application can handle.

**Testing Domain Errors:**

The goal when testing domain errors is to ensure that the application correctly identifies and responds to these errors as part of its normal operation.

> **Test samples from [FractionParserTest.kt](https://github.com/oppia/oppia-android/blob/develop/utility/src/test/java/org/oppia/android/util/math/FractionParserTest.kt):**

Reviewer: Let's make this a permalink so that it doesn't require updating if the file ever moves.

Author: Got it! updated it with a permalink.

```kotlin
@Test
fun testOrderPizza_withNoTakeaway_doesNotPrintTakeawayMessage() {
  val pizza = Pizza()
  val output = ByteArrayOutputStream()
  System.setOut(PrintStream(output))

  // Order pizza without takeaway, default value is false
  pizza.orderPizza(takeaway = false)

  // Verify that "with extra cheese" is not printed
  assertThat(output.toString().trim()).doesNotContain("for takeaway")

  System.setOut(System.out)
}
```

These tests confirm that the class behaves as expected without producing unnecessary outputs.

# Testing Public API

A public API (Application Programming Interface) refers to the set of methods, properties, and functionalities exposed by a class or module for use by external code. It defines how other parts of a system or external systems can interact with the functionality provided by that class or module.

Public APIs are essential because they provide a way to interact with the functionality of a class or module without exposing its internal workings. They define how external code can use the functionality offered by the class or module, ensuring that interactions are safe and predictable while keeping the internal implementation hidden and secure.

> Let's consider the following example for a public API to withdraw money from the BankAccount.

Reviewer: ```suggestion
Let's consider the following example for a public API to withdraw money from a BankAccount.
```

Nit on grammar.

Author: Thanks, updated as suggested.

    // Perform the withdrawal operation
    // Print a receipt if needed
    println("Withdrawing $amount for $requestedUsername")
    if (needReceipt) {
      printReceipt(amount)
    }

    // Process the file if provided
    file?.let {
      processFile(it)
    }
  }

  private fun isValidUser(requestedUsername: String, requestedPassword: String): Boolean {
    return true
  }

  private fun isValidWithdrawal(amount: Double): Boolean {
    return true
  }

  private fun performWithdrawal(amount: Double) {
    println("Withdrew $amount. New balance is $balance")
  }

  private fun printReceipt(amount: Double) {
    println("Receipt: Withdrew $amount. Current balance: $balance")
  }

  private fun processFile(file: File) {
    println("Processing file: ${file.name}")
  }
}
```

The **`withdraw`** method serves as the single public entry point for withdrawing money from the account. It handles user validation, amount checking, optional file upload and printing of the receipt. By keeping the internal methods private, the class ensures that the operations are performed in a controlled manner while hiding the complexity of these operations from the user.

> ## How to Write Tests for Public API

Reviewer: ```suggestion
## How to Write Tests for a Public API
```

Small readability adjustment.

Author: Done, updated as suggested.

> ```sh

Reviewer: Ditto here--should be unhighlighted code text since the block doesn't contain executable code.

Author: Replaced with mermaid graph.

> +---------+    +---------------+    +-----------+    +--------+

Reviewer: For this & the block above: perhaps these should be diagrams instead of plain text? Plain text may not consistently render across different browser/window sizes.

Author: Got it! Earlier kept the Quality vs Percentage illustration as text as graphs didn't properly fit the way I wanted them --

```mermaid

graph TD
    A[Coverage Analysis]
    B[100% Line Coverage<br>Does not guarantee full<br>behavioural coverage]
    C[Less Than 100% Coverage<br>Guarantees at least one<br>important behaviour is missing]
    D[Add tests for lines that<br>miss behavioural coverages<br>even if stated as covered]
    E[Find uncovered lines and add<br>tests that cover all behaviours<br>and not just line coverages]

    A --> B
    A --> C
    C --> D
    B ---> E
    D --> E

```

it works (didn't exactly fit in how I pictured it), but just found that there was another block element, that was close enough so replaced the Quality vs Percentage illustration with blocks 

```mermaid
block-beta
columns 2 
  A[["Coverage Analysis"]]:2
  block:Coverage1
    B["100% Line Coverage<br>Does not guarantee full<br>behavioural coverage"]
  end
  block:Coverage2
    C["Less Than 100% Coverage<br>Guarantees at least one<br>important behaviour is missing"]
  end
  block:Action1
    D["Add tests for lines that<br>miss behavioural coverages<br>even if stated as covered"]
  end
  block:Action2
    E["Find uncovered lines and add<br>tests that cover all behaviours<br>and not just line coverages"]
  end
  space
  Coverage1 --> D
  Coverage2 --> E
  E --> D

  style A fill:#f3f3f3,stroke:#333,stroke-width:1px
  style B fill:#e8fcde,stroke:#333,stroke-width:1px
  style C fill:#ffdad4,stroke:#333,stroke-width:1px
  style D fill:#d8fac5,stroke:#333,stroke-width:1px
  style E fill:#e2fcd4,stroke:#333,stroke-width:1px

```

and the API 

```mermaid
graph LR
    A[Analyze] --> B[Map Behaviors]
    B --> C[Add Tests]
    C --> D[Refine]

    style A fill:#f5f5f5,stroke:#333,stroke-width:1px
    style B fill:#f2fced,stroke:#333,stroke-width:1px
    style C fill:#e8fcde,stroke:#333,stroke-width:1px
    style D fill:#dcf5d0,stroke:#333,stroke-width:1px

```

and structuring test bodies illustrations with graphs 

```mermaid
graph LR
    A[&ensp;&nbsp;Tests&ensp;] -..->|verify| B[Source Code]
    C[Humans] -..->|verify| D[&emsp;&ensp;Tests&ensp;&emsp;]

    style A fill:#e8e8e8,stroke:#333,stroke-width:1px
    style B fill:#e0f7d7,stroke:#333,stroke-width:1px
    style C fill:#e8e8e8,stroke:#333,stroke-width:1px
    style D fill:#e0f7d7,stroke:#333,stroke-width:1px
```

also used the color code #e0f7d7 for mapping lines to code too as a common green code.

Reviewer: Ah, these look really good! I'm glad you were able to figure out a cleaner way to lay them out.

}

@Test
fun testWithdraw_withNeedReceipt_receiptPrinted() {
  val account = BankAccount(1000.0, "user", "password")
  
  val output = ByteArrayOutputStream()
  System.setOut(PrintStream(output))

  account.withdraw("user", "password", 200.0, needReceipt = true)

  assertThat(output.toString().trim()).contains("Receipt: Withdrew 200.0. Current balance: 800.0")
  System.setOut(System.out)
}

@Test
fun testWithdraw_withInvalidFileFormat_throwsInvalidFileFormatException() {
  val account = BankAccount(1000.0, "user", "password")
  val invalidFile = File("invalid.txt")

  val exception = assertThrows<InvalidFileFormatException> {
    account.withdraw("user", "password", invalidFile, 200.0)
  }
  assertThat(exception).hasMessageThat().contains("Invalid file format")
}

@Test
fun testWithdraw_withDefaultReceipt_noReceiptPrinted() {
  val account = BankAccount(1000.0, "user", "password")

  val output = ByteArrayOutputStream()
  System.setOut(PrintStream(output))

  account.withdraw("user", "password", 200.0)

  assertThat(output.toString().trim()).doesNotContain("Receipt:")
  System.setOut(System.out)
}

```

### 4. Review and Refine

> Goal: Ensure all scenarios are covered and tests are effective.

Reviewer: Should running coverage be a step here? It's possible that one missed something in the initial analysis and coverage would indicate that the list of behaviors needs to be reassessed based on coverage findings.

Author: Ah yeah, I was confined to testing and overlooked the code coverage process. Thanks, now added the iterative coverage analysis to the section.

  assertThat(output.toString().trim()).contains("New balance is 800.0")
  System.setOut(System.out)
}
```

**e. To verify with the balance is updated correctly:**

Ensures that the balance is updated correctly after a withdrawal.

```kotlin
@Test
fun testWithdraw_withSufficientBalance_updatesBalanceCorrectly() {
  account.withdraw("user", "password", 200.0)
  assertThat(account.balance).isEqualTo(800.0)
}
```

These tests cover various aspects of the withdrawal functionality, ensuring that the balance updates correctly with receipts, passbooks and output messages. Although the core functionality of withdrawing funds and updating the balance is consistent, it can be observed in multiple ways. Each test focuses on a specific verification method while ultimately validating the same core functionality.

# Structuring Test Functionalities

In testing, it's crucial to ensure that your tests verify implementation code while maintaining clarity and readability. Tests validate the correctness of the code, but it is humans who verify the correctness of the test code itself. Therefore, striking a balance between clarity and conciseness in test writing is essential.

> ```sh

Reviewer: Ditto here (for the Shell text highlighting and the note above about maybe diagramming this, instead).

Author: Done, replaced with graph representation.

+------------+     verify     +--------------+
|   Tests    |   --------->   |  Source Code |
+------------+                +--------------+

+------------+     verify     +--------------+

> |   Human    |   ---------->  |    Tests     |

Reviewer: Nit: 'humans' for consistency in plurality (with 'tests' above).

Author: Thanks, corrected it.

+------------+                +--------------+

```

> Tests should focus on verifying the behavior of the implementation code, while humans should be able to easily understand and verify the correctness of the test code itself.

Reviewer: Nit on 'behavior' we should probably use consistent American English or British English spelling everywhere.

Author: Ah, I doubted I was mixing them up. Did a full search and updated everything to 'behavior.

    discount = when {
      isMember && code == "SAVE10" -> 0.10
      code == "SAVE20" && quarter == YearQuarter.QUARTER1 -> 0.20
      code == "SUMMERSALE" && quarter == YearQuarter.QUARTER2 -> 0.15
      else -> 0.0
    }
    println("Discount applied: ${discount * 100}%")
  }

  private fun calculateBill(): Double {
    return items.sumOf { menu[it] ?: 0.0 }
  }
}

enum class YearQuarter {
    QUARTER1, QUARTER2, QUARTER3, QUARTER4
}
```

It's important to understand how to segment or split functionalities in your tests to maintain clarity and avoid confusion. Knowing when to use helper functions and `@Before` / `@After` annotations effectively, and when to keep logic within the test cases themselves, ensures that your tests are both clear and maintainable. Let’s explore these concepts with a Restaurant Ordering System example.

## 1. When and How to Divide Responsibilities

### a. Helper Functions

Helper Functions are valuable for reducing redundancy in tests. They encapsulate **non-behavioral tasks**, ensuring that the focus remains on testing the core logic.

**Helper Function:**

```kotlin
// Helper function to initialize a Restaurant with a menu

> fun createRestaurantWithMenu(): Restaurant {

Reviewer: This is actually a problematic helper because it actually hides aspects of the restaurant construction that could directly relate to the test body. For instance, 'burger' is essential for ``testPlaceOrder_withValidItems_displaysCorrectTotalBill`` below.

Author: Hmm... I thought that since they were provided as parameters and wouldn't be used in the core logic, they would only serve as default values for the menu (trying to replicate the initial setup). 

Now I've adapted the approach of sending test file exemptions as a reference and modified the Restaurant example it to accept a file path to store a order log history and using a utility function to load a text file for each test case Would this fit this situation?

Reviewer: I think the log file idea was really good--it's a much better demonstration of 'extra details.' The order does seem to be relevant since the tests are dealing with the concept of 'valid' vs. 'invalid' items (which directly ties to the orders).

  val restaurant = createRestaurantWithMenu()
  val discountDetails = createDiscount()

  restaurant.applyDiscount(discountDetails.first, discountDetails.second, discountDetails.third)
  restaurant.placeOrder(listOf("Burger"))

  assertThat((outputStream.toString().trim()).contains("Total bill after discount: 4.5")
}
```

**The Drawbacks of This Approach**
- Hidden Logic: The helper function `createDiscount()` hides critical logic affecting the test outcome. This makes the test harder to understand and debug.
- Complexity: The helper function handles multiple scenarios, which should be tested separately. This introduces complexity and reduces clarity.
- Clarity: Hidden logic in a helper function makes the test less transparent and harder to interpret, compromising its readability.

**Approach to write test with clarity:**

In test code, being explicit often trumps being concise. This means defining the necessary conditions and actions directly within the test case, so the test's intent is immediately clear to anyone reading it.

```kotlin
@Test
fun testDiscountedBill_withAppliedDicount_returnsDiscountedBill() {

>   val restaurant = createRestaurantWithMenu()

Reviewer: Ditto here--solution for creating the restaurant above should also be done here.

Author: Now the restaurant declarations are done right in the test cases while helper functions are replaced to use the file creation process.

    
  // Explicitly defining discount details in the test
  val isMember = true
  val code = "SAVE10"
  val quarter = YearQuarter.QUARTER1

  restaurant.applyDiscount(isMember, code, quarter)
  restaurant.placeOrder(listOf("Burger"))

  assertThat((outputStream.toString().trim()).contains("Total bill after discount: 4.5")
}
```

Laying out the logic and conditions directly within the test makes it independent of external functions or files, which makes the test easier to understand, maintain, and debug.

Unlike production code, where duplication is often avoided, in test code, it’s sometimes better to duplicate code if it leads to clearer, more understandable tests. This ensures that the behavior being tested is directly represented in the test case.

## 3. Importance of Descriptive Test Names

Naming test functions descriptively helps in identifying the purpose and scope of each test. Use names that reflect the specific behavior being tested.

> Oppia Android follows a naming convention where the test names should read like a sentence, and be consistent with other nearby test names to facilitate easily coming up with new tests. Consider using a format similar to the following for naming test functions:

Reviewer: ```suggestion
Oppia Android follows a naming convention where the test names should read like a sentence, and be consistent with other nearby test names to facilitate easily coming up with new tests. It's preferred that the following format be used for naming test functions:
```

Strengthening language since we actually want people to use this format.

Author: Done, updated as suggested.

```
testAction_withOneCondition_withSecondCondition_hasExpectedOutcome
```

```kotlin
@Test
fun testPlaceOrder_withValidItems_orderPlacedSuccessfully() {
  // Test Logic: Order should be placed with given items
}

@Test
fun testPlaceOrder_withEmptyItems_orderNotPlaced() {
  // Test Logic: Handle empty order gracefully
}

@Test
fun testCalculateBill_withValidItems_correctBillCalculated() {
  // Test Logic: Calculate correct bill for ordered items
}

@Test
fun testApplyDiscount_withMemberAndValidCode_discountApplied() {
  // Test Logic: Apply 10% discount for valid code and membership
}

@Test
fun testApplyDiscount_withNonMemberAndValidCode_noDiscountApplied() {
  // Test Logic: No discount applied for non-member
}

@Test

> fun testApplyDiscount_withMemberAndQuarter2_discountApplied() {

Reviewer: Nit: none of the examples show the second condition with an underscore--perhaps change one of the examples to match that multi-condition case more explicitly?

Author: Updated it to `testApplyDiscount_withMemberAndValidCode_inQuarter2_discountApplied`, would this meet the requirement or should it be split for each condition as `testApplyDiscount_withMember_validCode_inQuarter2_discountApplied`, `testApplyDiscount_withNonMember_validCode_noDiscountApplied` .

Reviewer: I think the new one is sufficient for demonstrating the example. It could, as you pointed out, be further split up or refined, but we don't enforce that strict of a standard (I was hoping more for a clear example than a set of definitions/rules).
